From 3b7a430088ee269c53bb3d107ad13da23e795a7b Mon Sep 17 00:00:00 2001
From: Caiwen Zhang <caiwen.zhang@intel.com>
Date: Fri, 28 Sep 2012 18:31:58 +0800
Subject: [PATCH 12/23] Fix EFsmsp size error
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

EFsmsp record size 28 + Y (Y is the bytes remained for saving alpha
id). Y always be (record size - 28), it doesn't depends on the alpha
string length, if the alpha string is longer than Y, the string should
be cut short, if it is shorter than Y, should padding with 0xFF.
---
 src/s_sms.c |   15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/src/s_sms.c b/src/s_sms.c
index 2b3cd4f..20067a3 100644
--- a/src/s_sms.c
+++ b/src/s_sms.c
@@ -3046,6 +3046,7 @@ static TReturn set_sms_params(CoreObject *obj, UserRequest *ur)
 	char *encoded_data = NULL;
 	unsigned char *temp_data = NULL;
 	int SMSPRecordLen = 0;
+	int *smsp_record_len;
 
 	TcoreHal *hal = NULL;
 	TcoreATRequest *atreq = NULL;
@@ -3067,11 +3068,15 @@ static TReturn set_sms_params(CoreObject *obj, UserRequest *ur)
 		return TCORE_RETURN_ENOSYS;
 	}
 
-	//EFsmsp file size is 28 +Y bytes (Y is alpha id size)
-	SMSPRecordLen = 28 + setSmsParams->params.alphaIdLen;
-	temp_data = calloc(SMSPRecordLen,1);
-	encoded_data = calloc(SMSPRecordLen*2 + 1,1);
-	
+	// EFsmsp file size is 28 +Y bytes (Y is alpha id size)
+	smsp_record_len = tcore_plugin_ref_property(tcore_object_ref_plugin(obj), "SMSPRECORDLEN");
+	SMSPRecordLen = *smsp_record_len;
+	if (SMSPRecordLen < nDefaultSMSPWithoutAlphaId)
+		return FALSE;
+
+	temp_data = calloc(SMSPRecordLen, 1);
+	encoded_data = calloc(SMSPRecordLen * 2 + 1, 1);
+
 	_tcore_util_sms_encode_smsParameters(&(setSmsParams->params), temp_data, SMSPRecordLen);
 
 	util_byte_to_hex((const char *)temp_data, (char *)encoded_data,SMSPRecordLen);
-- 
1.7.10.4

